soc: stm32u0: add PM

mickael868 opened the pull request:
This PR implements the Power Management (PM) for the STM32U0x targets. It is mostly based on the STM32U5 target PM implementation. It is a follow up of the following [aborted PR](https://github.com/zephyrproject-rtos/zephyr/pull/88878).

Tested with the following code:

```c
#include <zephyr/kernel.h>

int main()
{
    struct k_event event;
    k_event_init(&event);
    while (1)
    {
        k_event_wait(&event, 1, false, K_FOREVER);
    }
    return 0;
}
```

When `CONFIG_PM*` is set to `y`, the power consumption is about 80 µA when nothing is connected to the target. The result is the same with and without the following overlay:

```c
&clk_lsi {
  status = "okay";
};

stm32_lp_tick_source: &lptim2 {
  clocks = <&rcc STM32_CLOCK(APB1, 30U)>,
           <&rcc STM32_SRC_LSI LPTIM2_SEL(1)>;
  // st,prescaler = <8>; //doe not reduce power consumption
  status = "okay";
};
```

When `CONFIG_PM*` flags are commented the power consumption is about 3 mA.

It is necessary to disable the `pm_policy_state_lock_get()` calls from `drivers/serial/uart_stm32.c` because it takes the lock once at the first `printk()` call and never put it back. This is the get backtrace if the boot banner is enabled:

```
Breakpoint 1, uart_poll_out (dev=<optimized out>, out_char=42 '*') at app/build/zephyr/include/generated/zephyr/syscalls/uart.h:108
108             z_impl_uart_poll_out(dev, out_char);
(gdb) backtrace
#0  uart_poll_out (dev=<optimized out>, out_char=42 '*') at app/build/zephyr/include/generated/zephyr/syscalls/uart.h:108
#1  console_out (c=42) at app/zephyr/drivers/console/uart_console.c:102
#2  0x080004a4 in char_out (c=<optimized out>, ctx_p=<optimized out>) at app/zephyr/lib/os/printk.c:98
#3  0x08004216 in vfprintf ()
#4  0x080004de in vprintk (fmt=0x8007689 "*** Booting Zephyr OS build v4.1.0-2646-ge41909a32ce9 ***\n", fmt@entry=0x1 <error: Cannot access memory at address 0x1>, ap=..., ap@entry=...) at app/zephyr/lib/os/printk.c:134
#5  0x0800560c in printk (fmt=0x8007689 "*** Booting Zephyr OS build v4.1.0-2646-ge41909a32ce9 ***\n") at app/zephyr/lib/os/printk.c:198
#6  0x08003ff0 in boot_banner () at app/zephyr/kernel/banner.c:46
#7  0x08002cc0 in bg_thread_main (unused1=<optimized out>, unused2=<optimized out>, unused3=<optimized out>) at app/zephyr/kernel/init.c:551
#8  0x08000508 in z_thread_entry (entry=0x8002cad <bg_thread_main>, p1=0x0, p2=0x0, p3=0x0) at app/zephyr/lib/os/thread_entry.c:48
#9  0x0800108c in arch_switch_to_main_thread (main_thread=0x8002cad <bg_thread_main>, stack_ptr=0x0, _main=0x8002cad <bg_thread_main>) at /home/mika/dev/embedded/zephyr/common/zephyr/arch/arm/core/cortex_m/thread.c:546
#10 0x20000608 in z_idle_threads ()
```

`uart_stm32_pm_policy_state_lock_put_unconditional` is never called. Tested with `CONFIG_UART_INTERRUPT_DRIVEN=y`.

I'm quiet new to Zephyr, does the UART issue comes an error I made?
Also, I think we should expect a lower power consumption when the LPTIM2 timer is used, do you have any idea how to improve this using only internal clocks? I will try to add a LSE oscillator in the next design (mainly for the LPUARTx peripherals) but I can not test it for now.

Thanks.

mickael868 commented:
> Please open a bug issue for the UART PM problem.

Done [here](https://github.com/zephyrproject-rtos/zephyr/issues/89056)

etienne-lms commented on soc/st/stm32/stm32u0x/soc.c:
```diff
@@ -20,6 +20,7 @@
 #define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
 LOG_MODULE_REGISTER(soc);
 
+extern void stm32_power_init(void);
```
I would suggestion to move this declaration in local **soc.h**.

mickael868 replied:
I did apply the suggestion, enclosing the declaration in a `#ifdef CONFIG_PM`.

Should I revert?

etienne-lms commented on soc/st/stm32/stm32u0x/CMakeLists.txt:
```diff
@@ -2,6 +2,7 @@
 
 zephyr_include_directories(${ZEPHYR_BASE}/drivers)
 zephyr_sources(
+  power.c
```
`zephyr_sources_ifdef(CONFIG_PM power.c)`?

etienne-lms reviewed (commented):
@mickael868, thanks for the updates.
Could you squash the fixup commits into the 1st commit?

mickael868 commented on soc/st/stm32/stm32u0x/soc.c:
```diff
@@ -35,4 +39,8 @@ void soc_early_init_hook(void)
 	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
+
+#ifdef CONFIG_PM
+	stm32_power_init();
```
No problem, I should have checked this too.

mickael868 commented on soc/st/stm32/stm32u0x/power.c:
```diff
@@ -0,0 +1,114 @@
+#include <zephyr/kernel.h>
+#include <zephyr/pm/pm.h>
+#include <soc.h>
+#include <zephyr/init.h>
```
Done by removing the unused `soc.h` include and applying `clang-format` with the options `IncludeBlocks` and `SortIncludes` enabled.

erwango commented on drivers/timer/stm32_lptim_timer.c:
Please make changes to this file a dedicated commit.

erwango reviewed (commented):
Otherwise LGTM

erwango commented on soc/st/stm32/stm32u0x/power.c:
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright The Zephyr Project Contributors
```
FYI (not blocking): This isn't wrong, but note that you could also use your own copyright.

erwango commented on soc/st/stm32/stm32u0x/power.c:
```diff
@@ -0,0 +1,107 @@
+#include <zephyr/pm/pm.h>
+
+#include <clock_control/clock_stm32_ll_common.h>
+#include <stm32u0xx_ll_bus.h>
+#include <stm32u0xx_ll_cortex.h>
+#include <stm32u0xx_ll_pwr.h>
+#include <stm32u0xx_ll_rcc.h>
+#include <stm32u0xx_ll_system.h>
+#include <stm32u0xx_ll_utils.h>
```
Last minute catch:
Since only LL_PWR, LL_LPM and LL_RCC are used in this file, I'm pretty sure at least 2 (bus, cortex) of those are not required (maybe 3).

erwango reviewed (changes requested):
Otherwise LGTM

Commit:
soc: stm32u0: add PM

STM32U0 specific changes to enable the PM feature.
Base on the power-related code from the STM32U5 target.

Signed-off-by: Mickael Bosch <mickael.bosch@linux.com>

Commit:
drivers: timer: handle the stm32u0 target

Configure the lptim to wake-up from sleep.

Signed-off-by: Mickael Bosch <mickael.bosch@linux.com>

kartben commented:
@mickael868 please follow-up on @mathieuchopstm's comment (alternatively, @mathieuchopstm, please consider approving the PR if your comments are non-blocking)
Thanks!

mathieuchopstm reviewed (approved):
My nits should be non-blocking
